src/main/java/com/google/cloud/tools/appengine/api/deploy/DefaultStageStandardConfiguration.java

```diff
@@ -35,123 +39,165 @@
   @Nullable
   public Boolean getDisableJarJsps() {
     return disableJarJsps;
   }
 
-  public void setDisableJarJsps(Boolean disableJarJsps) {
-    this.disableJarJsps = disableJarJsps;
-  }
-
   @Override
   @Nullable
   public String getRuntime() {
     return runtime;
   }
 
-  public void setRuntime(String runtime) {
-    this.runtime = runtime;
+  public static class Builder {
+
+    @Nullable private File sourceDirectory;
+    @Nullable private File stagingDirectory;
+    @Nullable private File dockerfile;
+    @Nullable private Boolean enableQuickstart;
+    @Nullable private Boolean disableUpdateCheck;
+    @Nullable private Boolean enableJarSplitting;
+    @Nullable private String jarSplittingExcludes;
+    @Nullable private String compileEncoding;
+    @Nullable private Boolean deleteJsps;
+    @Nullable private Boolean enableJarClasses;
+    @Nullable private Boolean disableJarJsps;
+    @Nullable private String runtime;
+
+    public Builder setSourceDirectory(File sourceDirectory) {
+      this.sourceDirectory = Preconditions.checkNotNull(sourceDirectory);
+      return this;
+    }
+
+    public Builder setStagingDirectory(File stagingDirectory) {
+      this.stagingDirectory = Preconditions.checkNotNull(stagingDirectory);
+      return this;
+    }
+
+    public void setDockerfile(@Nullable File dockerfile) {
+      this.dockerfile = dockerfile;
+    }
+
+    public void setEnableQuickstart(@Nullable Boolean enableQuickstart) {
+      this.enableQuickstart = enableQuickstart;
+    }
+
+    public void setDisableUpdateCheck(@Nullable Boolean disableUpdateCheck) {
+      this.disableUpdateCheck = disableUpdateCheck;
+    }
+
+    public void setEnableJarSplitting(@Nullable Boolean enableJarSplitting) {
+      this.enableJarSplitting = enableJarSplitting;
+    }
+
+    public void setJarSplittingExcludes(@Nullable String jarSplittingExcludes) {
+      this.jarSplittingExcludes = jarSplittingExcludes;
+    }
+
+    public void setCompileEncoding(@Nullable String compileEncoding) {
+      this.compileEncoding = compileEncoding;
+    }
+
+    public void setDeleteJsps(@Nullable Boolean deleteJsps) {
+      this.deleteJsps = deleteJsps;
+    }
+
+    public void setEnableJarClasses(@Nullable Boolean enableJarClasses) {
+      this.enableJarClasses = enableJarClasses;
+    }
+
+    public void setDisableJarJsps(@Nullable Boolean disableJarJsps) {
+      this.disableJarJsps = disableJarJsps;
+    }
+
+    public void setRuntime(@Nullable String runtime) {
+      this.runtime = runtime;
+    }
+
+    /** Builds the configuration. */
+    public StageStandardConfiguration build() {
+      if (sourceDirectory == null || stagingDirectory == null) {
+        throw new NullPointerException("Incomplete configuration");
+      }
+      DefaultStageStandardConfiguration stageStandardConfiguration =
+          new DefaultStageStandardConfiguration(sourceDirectory, stagingDirectory);
+
+      stageStandardConfiguration.dockerfile = dockerfile;
```

Comment:
I think the pattern, in general, calls for the members of the class to be final, and everything passed through the contructor (as nullable?) Instead of setting them here via scope privilege.

Reply:
That is one way to do it, but it's a private implementation detail, not a fundamental feature of builders. As long as a reference to the object doesn't escape before the method returns, there's no reason not to do this. There's also one strong reason to do this. This approach is far less susceptible to accidentally swapping arguments. If we just passed them all into a constructor, would you notice if  jarSplittingExcludes was being passed where deleteJsps was expected and vice versa? Private or not, very long argument lists of the same type are risky.

Reply:
I mean you would probably catch that testing your builder. It's just a little strange to me to see it like this.

Reply:
Or pass the builder instance into the constructor?

Reply:
I'm not confident enough in the existing test coverage to rely on it exclusively.
Passing the builder instance into the constructor is an interesting trick I hadn't thought of. It seems to flip flop the existing approach.